core/templates/dev/head/pages/exploration-editor-page/feedback-tab/services/thread-data.service.ts

```diff
@@ -50,25 +50,23 @@ angular.module('oppia').factory('ThreadDataService', [
     // All the threads for this exploration. This is a list whose entries are
     // objects, each representing threads. The 'messages' key of this object
     // is updated lazily.
-    var _data = {
-      feedbackThreads: [],
-      suggestionThreads: []
-    };
+    var _data = { feedbackThreads: [], suggestionThreads: [] };
+    var threadsById = {};
 
-    // TODO(brianrodri@): Use a helper-object to give this function O(1)
-    // complexity instead of O(N).
-    var getThreadById = function(threadId) {
-      var thread = null;
-      var allThreads = _data.feedbackThreads.concat(_data.suggestionThreads);
-      for (var i = 0; i < allThreads.length; i++) {
-        if (allThreads[i].threadId === threadId) {
-          thread = allThreads[i];
-          break;
-        }
+    var setThread = function(thread) {
+      var oldThread = threadsById[thread.threadId];
+      if (oldThread) {
+        thread = oldThread.copy(thread);
+      } else {
+        threadsById[thread.threadId] = thread;
       }
       return thread;
     };
 
+    var getThreadById = function(threadId) {
+      return threadsById[threadId];
```

Comment:
Should there be a defensive assertion here that the thread ID exists in the dict?